### DISCUSSION TITLE
Best way to display images from private bucket?

### DISCUSSION THREAD

**Discussion by pierroo** [supabase/supabase#22050] [Questions] [answered]
Hey guys,

I am struggling to find the best (and safest) way to display an image in react native, from a private bucket.

I have rules on that bucket that for example I can only see the folder name with my own auth.uid()

I can safely upload to it.
Obviously when from the upload I then "getPublicUrl" => that URL just won't display anything as a source in an <Image> from RN.

So far I see two ways to solve it:

1- I get the signed URL instead, with a suuuuuper long expiration date (10 years)
and then I save that signed url in my database
But then that signed URL, anyone can then see it, no? (obviously, they would still need to know that URL)
is that safe for example in a chat app where users share private photos?

2- I tried to send the authorization in the <Image headers, like so:
```
<Image
        source={{
          uri: photo,
          cache: 'force-cache'
        }}
        headers={{
          'apikey' : process.env.SUPABASE_ANON_KEY,
          'Authorization' : 'Bearer ' + session.access_token
        }}
        style={{ width: size, height: imageHeight, resizeMode: 'contain' }}
      />
```
that one unfortunately doesn't display anything, although I was hoping the headers props would work with my session token.

So, what is the correct way to display private images from a private bucket?

**Comment by GaryAustin1** [comment, marked as answer]
Signed URL is the proper way.  You would generate them from the path stored in the database when needed, not store the signed URL.  

If you are willing to do a 10yr expire, then just do a public bucket with hard to guess (UUID) filenames.  You would need to know the url to use the file.  If the user gives it away it can be accessed, but that is not different than a signedURL with any time longer than minutes.  
Remember a public bucket ONLY provides the ability for a public URL to read the file.  All other operation (upload, list, etc) all require RLS.

**Reply by pierroo** [reply]
Thank you for your answer.
Unfortunately "generating a signed URL each time they are needed" is not manageable, when we are talking about displaying a bunch of chat photos, with a chat that is opened often and that relies on <Image with source uri etc.

I will then go with the public bucket, reassuring to know that no one can list what's inside beyond RLS.

EDIT: so you confirm it's impossible for anyone to even list the folder names (which will also be uuid, on top of file names containing uuids) inside the buckets? (beyond RLS, again)
